server/src/main/java/org/elasticsearch/cluster/service/MasterService.java

```diff
@@ -1425,18 +1425,16 @@ void runBatch(
         );
     }
 
-    private static class TaskTimeoutHandler extends AbstractRunnable {
+    private static class TaskTimeoutHandler<T extends ClusterStateTaskListener> extends AbstractRunnable {
 
         private final TimeValue timeout;
         private final String source;
-        private final AtomicBoolean executed;
-        private final ClusterStateTaskListener listener;
+        private final AtomicReference<T> taskHolder;
```

Comment:
Maybe a comment here would be nice to explain the state. Something like

> Holds the task until either it's got for execution, or it times out, whichever comes first, and is then set to null.

Reply:
I added a comment in fbc35deb6d4.